<a href="https://colab.research.google.com/github/SriRamK345/Recommendation_System/blob/main/Recommendation_sys.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [ ]:
credit_df = pd.read_csv("/content/drive/MyDrive/Recommendation/credits.csv")
movies_df = pd.read_csv("/content/drive/MyDrive/Recommendation/movies.csv")

In [ ]:
display(credit_df.head())
movies_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   10-12-2009  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   19-05-2007   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   26-10-2015   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   16-07-2012  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   07-03-2012   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [ ]:
print("Credits:",credit_df.shape)
print("Movies Dataframe:",movies_df.shape)

Credits: (4803, 4)
Movies Dataframe: (4803, 20)


In [ ]:
df = movies_df.merge(credit_df, on='title')
df.sample(6)

budget                                             genres  \
4735         0  [{"id": 18, "name": "Drama"}, {"id": 35, "name...   
1968  25000000  [{"id": 28, "name": "Action"}, {"id": 18, "nam...   
3052  10000000                      [{"id": 18, "name": "Drama"}]   
2541  15000000  [{"id": 35, "name": "Comedy"}, {"id": 10751, "...   
1716  21000000  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
2088  21000000                      [{"id": 18, "name": "Drama"}]   

                                       homepage      id  \
4735                                        NaN  175291   
1968  https://www.facebook.com/JaneGotAGunFilm/  174751   
3052                                        NaN   87093   
2541                                        NaN   11870   
1716                                        NaN    9428   
2088            http://www.thebeaver-movie.com/   50780   

                                               keywords original_language  \
4735  [{"id": 9937, "name": "suspense"}, {"id": 2100...                en   
1968                                                 []                en   
3052  [{"id": 1157, "name": "wife husband relationsh...                en   
2541  [{"id": 776, "name": "screenplay"}, {"id": 887...                en   
1716  [{"id": 3446, "name": "forgiveness"}, {"id": 5...                en   
2088  [{"id": 6300, "name": "puppet"}, {"id": 187056...                en   

            original_title                                           overview  \
4735         Cheap Thrills  Recently fired and facing eviction, a new dad ...   
1968        Jane Got a Gun  After her outlaw husband returns home shot wit...   
3052              Big Eyes  The story of the awakening of the painter, Mar...   
2541          Big Fat Liar  Fourteen-year-old Jason Shepherd (Muniz) has a...   
1716  The Royal Tenenbaums  An estranged family of former child prodigies ...   
2088            The Beaver  Suffering from a severe case of depression, to...   

      popularity                               production_companies  ...  \
4735   10.826253  [{"name": "New Artists Alliance", "id": 5124},...  ...   
1968   15.333575  [{"name": "Handsomecharlie Films", "id": 8873}...  ...   
3052   28.570174  [{"name": "The Weinstein Company", "id": 308},...  ...   
2541    9.125594         [{"name": "Universal Pictures", "id": 33}]  ...   
1716   24.130810      [{"name": "Touchstone Pictures", "id": 9195}]  ...   
2088   13.847872  [{"name": "Summit Entertainment", "id": 491}, ...  ...   

     runtime                                   spoken_languages    status  \
4735    85.0           [{"iso_639_1": "en", "name": "English"}]  Released   
1968    98.0           [{"iso_639_1": "en", "name": "English"}]  Released   
3052   105.0  [{"iso_639_1": "it", "name": "Italiano"}, {"is...  Released   
2541    88.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1716   110.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
2088    91.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                              tagline                 title  \
4735             Get ready for the night of your life         Cheap Thrills   
1968                                              NaN        Jane Got a Gun   
3052  She created it. He sold it. And they bought it.              Big Eyes   
2541                     The truth is never overrated          Big Fat Liar   
1716            Family isn’t a word, it’s a sentence.  The Royal Tenenbaums   
2088                 He's here to save Walter's life.            The Beaver   

     vote_average vote_count movie_id  \
4735          6.3        155   175291   
1968          5.4        285   174751   
3052          6.8       1196    87093   
2541          5.6        197    11870   
1716          7.4       1286     9428   
2088          6.3        276    50780   

                                                   cast  \
4735  [{"cast_id": 2, "chara

In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4808 entries, 0 to 4807
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4808 non-null   int64  
 1   genres                4808 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4808 non-null   int64  
 4   keywords              4808 non-null   object 
 5   original_language     4808 non-null   object 
 6   original_title        4808 non-null   object 
 7   overview              4805 non-null   object 
 8   popularity            4808 non-null   float64
 9   production_companies  4808 non-null   object 
 10  production_countries  4808 non-null   object 
 11  release_date          4807 non-null   object 
 12  revenue               4808 non-null   int64  
 13  runtime               4806 non-null   float64
 14  spoken_languages      4808 non-null   object 
 15  status               

In [ ]:
print(df.shape)

(4808, 23)


In [ ]:
df.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count', 'movie_id', 'cast', 'crew'],
      dtype='object')

In [ ]:
df = df.loc[:,["movie_id","title","overview","genres","keywords","cast","crew"]]
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [ ]:
df.genres[0], print(type(df.genres[0]))

<class 'str'>


('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]',
 None)

In [ ]:
import ast
ast.literal_eval(df.genres[0])

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [ ]:
df.columns

Index(['movie_id', 'title', 'overview', 'genres', 'keywords', 'cast', 'crew'], dtype='object')

In [ ]:
df.overview[0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [ ]:
df.keywords[0]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [ ]:
df.cast[0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [ ]:
df.crew[0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [ ]:
df.shape

(4808, 7)

In [ ]:
df.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

a=[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]
for i in a:
  print(i["name"])

Action
Adventure
Fantasy
Science Fiction

In [ ]:
df.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

In [ ]:
import ast  # uesd to remove string
ast.literal_eval(df.iloc[0].genres)

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [ ]:
for i in ast.literal_eval(df.iloc[0].genres):
      print(i)

{'id': 28, 'name': 'Action'}
{'id': 12, 'name': 'Adventure'}
{'id': 14, 'name': 'Fantasy'}
{'id': 878, 'name': 'Science Fiction'}


In [ ]:
import ast
def convert1(obj):
  l = []
  for i in ast.literal_eval(obj):   #[{}{}{}]
    l.append(i['name'])
  return l

def convert2(obj):
  l = []
  counter = 0
  for i in ast.literal_eval(obj):
    if counter == 3:
      break
    l.append(i['name'])
    counter = counter + 1
  return l

def convert3(obj):
  l = []
  for i in ast.literal_eval(obj):
    if i['job'] == "Director":
      l.append(i['name'])
  return l

In [ ]:
df['genres']= df['genres'].apply(convert1)
df['keywords']= df['keywords'].apply(convert1)
df['cast'] = df['cast'].apply(convert2)
df['crew'] = df['crew'].apply(convert3)
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

In [ ]:
type(df['overview'][0])

str

In [ ]:
# df['overview'] = df['overview'].apply(lambda x : x.split())
# Handle potential float values by converting them to strings first
df['overview'] = df['overview'].astype(str).apply(lambda x: x.split())

In [ ]:
df.sample(2)

movie_id               title  \
3655     10844     Out of the Blue   
3951     22488  Love's Abiding Joy   

                                               overview  \
3655  [Ordinary, people, find, extraordinary, courag...   
3951  [The, continued, Westward, journey, of, settle...   

                                 genres  \
3655                            [Drama]   
3951  [TV Movie, Action, Drama, Family]   

                                               keywords  \
3655  [new zealand, gun rampage, police, independent...   
3951                                                 []   

                                                   cast                  crew  
3655            [Karl Urban, Tandi Wright, Simon Ferry]      [Robert Sarkies]  
3951  [Erin Cottrell, Logan Bartholomew, William Mor...  [Michael Landon Jr.]

In [ ]:
df['genres']= df['genres'].apply(lambda x : [i.replace(" ","") for i in x])
df['keywords']= df['keywords'].apply(lambda x : [i.replace(" ","") for i in x])
df['cast'] = df['cast'].apply(lambda x : [i.replace(" ","") for i in x])
df['crew'] = df['crew'].apply(lambda x : [i.replace(" ","") for i in x])

In [ ]:
df['tags']= df['overview']  + df['keywords'] + df['cast'] + df['crew'] + df['genres']
df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [ ]:
new_df = df[['movie_id','title','tags']]
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [ ]:
new_df["tags"][0]

['In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.',
 'cultureclash',
 'future',
 'spacewar',
 'spacecolony',
 'society',
 'spacetravel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alienplanet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'loveaffair',
 'antiwar',
 'powerrelations',
 'mindandsoul',
 '3d',
 'SamWorthington',
 'ZoeSaldana',
 'SigourneyWeaver',
 'JamesCameron',
 'Action',
 'Adventure',
 'Fantasy',
 'ScienceFiction']

In [ ]:
new_df['tags'] = new_df['tags'].apply(lambda x : " ".join(x))
new_df

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4803      9367                               El Mariachi   
4804     72766                                 Newlyweds   
4805    231617                 Signed, Sealed, Delivered   
4806    126186                          Shanghai Calling   
4807     25975                         My Date with Drew   

                                                   tags  
0     In the 22nd century, a paraplegic Marine is di...  
1     Captain Barbossa, long believed to be dead, ha...  
2     A cryptic message from Bond’s past sends him o...  
3     Following the death of District Attorney Harve...  
4     John Carter is a war-weary, former military ca...  
...                                                 ...  
4803  El Mariachi just wants to play his guitar and ...  
4804  A newlywed couple's honeymoon is upended by th...  
4805  "Signed, Sealed, Delivered" introduces a dedic...  
4806  When ambitious New York attorney Sam is sent t...  
4807  Ever since the second grade when he first saw ...  

[4808 rows x 3 columns]

In [ ]:
new_df["tags"][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron Action Adventure Fantasy ScienceFiction'

In [ ]:
import nltk
from nltk.stem.porter import PorterStemmer

ps = PorterStemmer()


In [ ]:
def stem(text):
  y = []
  for i in text.split():
    y.append(ps.stem(i))
  return " ".join(y)

In [ ]:
new_df['tags'] = new_df['tags'].apply(stem)

In [ ]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  in the 22nd century, a parapleg marin is dispa...  
1  captain barbossa, long believ to be dead, ha c...  
2  a cryptic messag from bond’ past send him on a...  
3  follow the death of district attorney harvey d...  
4  john carter is a war-weary, former militari ca...

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=5000, stop_words="english")

In [ ]:
cv.fit_transform(new_df['tags']).toarray().shape

(4808, 5000)

In [ ]:
cv.fit_transform(new_df['tags']).toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [ ]:
vectors = cv.fit_transform(new_df['tags']).toarray()

In [ ]:
vectors[0]

array([0, 0, 0, ..., 0, 0, 0])

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
cosine_similarity(vectors)

array([[1.        , 0.08458258, 0.08718573, ..., 0.04559608, 0.        ,
        0.        ],
       [0.08458258, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
        0.02615329],
       [0.08718573, 0.06063391, 1.        , ..., 0.02451452, 0.        ,
        0.        ],
       ...,
       [0.04559608, 0.02378257, 0.02451452, ..., 1.        , 0.03962144,
        0.04229549],
       [0.        , 0.        , 0.        , ..., 0.03962144, 1.        ,
        0.08714204],
       [0.        , 0.02615329, 0.        , ..., 0.04229549, 0.08714204,
        1.        ]])

In [ ]:
cosine_similarity(vectors).shape

(4808, 4808)

In [ ]:
similarity = cosine_similarity(vectors)

In [ ]:
list(similarity[0])

[0.9999999999999998,
 0.08458258116519013,
 0.08718572905786445,
 0.074458079104994,
 0.19184045508446734,
 0.11083777463036817,
 0.04078236951430929,
 0.1487044791289829,
 0.06003002251876642,
 0.09802861627917438,
 0.10397504898200727,
 0.09592022754223367,
 0.09158437663089952,
 0.04559607525875532,
 0.12996881122750908,
 0.06367145399670132,
 0.08000711205939975,
 0.14165281524036175,
 0.09620723036050562,
 0.08419653359482289,
 0.059053859535749265,
 0.11115400287534567,
 0.06711560552140244,
 0.0885807893036239,
 0.05405405405405406,
 0.03465834966066909,
 0.1559625734730109,
 0.18944220058835154,
 0.11810771907149853,
 0.06590621627456203,
 0.06774581282657174,
 0.16120647005479025,
 0.0863465783552691,
 0.09863939238321438,
 0.0,
 0.10067340828210367,
 0.17348855898381338,
 0.08000711205939975,
 0.08219949365267865,
 0.08336550215650926,
 0.07798128673650545,
 0.1678603981989089,
 0.0,
 0.09208185110322616,
 0.03427955850955305,
 0.08284930213349333,
 0.14097096860865022,
 0.19

In [ ]:
sorted(list(enumerate(similarity[0])),reverse = True,key=lambda x : x[1] )[1:6]

[(1216, 0.29061909685954823),
 (3731, 0.26401000024165),
 (507, 0.25903973506580724),
 (539, 0.2537477434955704),
 (2409, 0.2507061052819501)]

In [ ]:
def recommend(movie):
  movie_index = new_df[new_df["title"] == movie].index[0]
  distances = similarity[movie_index]
  movies_list = sorted(list(enumerate(distances)),reverse = True, key=lambda x : x[1])[1:6]

  for i in movies_list:
    print(new_df.iloc[i[0]].title)

In [ ]:
movie_index = new_df[new_df["title"] == "Pirates of the Caribbean: At World's End"].index[0]
movie_index

1

In [ ]:
distances = similarity[movie_index]

In [ ]:
movies_list = sorted(list(enumerate(distances)),reverse = True, key=lambda x : x[1])[1:6]
movies_list

[(12, 0.4140043409440133),
 (199, 0.27500954910846337),
 (17, 0.2533201985524494),
 (216, 0.20579830217101058),
 (3573, 0.20579830217101058)]

In [ ]:
for i in movies_list:
    print(new_df.iloc[i[0]].title)

Pirates of the Caribbean: Dead Man's Chest
Pirates of the Caribbean: The Curse of the Black Pearl
Pirates of the Caribbean: On Stranger Tides
Life of Pi
20,000 Leagues Under the Sea


In [ ]:
recommend("Pirates of the Caribbean: At World's End")

Pirates of the Caribbean: Dead Man's Chest
Pirates of the Caribbean: The Curse of the Black Pearl
Pirates of the Caribbean: On Stranger Tides
Life of Pi
20,000 Leagues Under the Sea


In [ ]:
a=[5,6,8,9]
for i,j in enumerate(a):
  print(i,j)

0 5
1 6
2 8
3 9


In [ ]:
sorted([9,6,8,9])

[6, 8, 9, 9]

In [ ]:
sorted(enumerate([9,6,8,78]),reverse = True,key=lambda x : x[1])[1:3]

[(0, 9), (2, 8)]

In [ ]:
def models(x_train,y_train):
  dic={}
  regressor = DecisionTreeRegressor(random_state = 1)


  Model=regressor.fit(X_train, y_train)
  y_pred = Model.predict(X_test)